### DISCUSSION TITLE
CR-SQLite extension and SQLite commit/rollback/change hooks

### DISCUSSION THREAD

**Discussion by michaelmedellin** [groue/GRDB.swift#1608] [General]
I'm running Custom GRDB with the [CR-SQLite](https://github.com/vlcn-io/cr-sqlite) extension loaded successfully. One of the features of CR-SQLite is that it increments a database version on every transaction commit to the database. For some reason, this is not happening correctly with GRDB. 

The _only_ time the database version in crsqlite will increment is once the connection to the database is closed. A new connection will increase the db version and apply that same version to all transactions until the connection is closed again. 

I've tried using both DatabaseQueue and DatabasePool without success. I did get the version to increment after manually checkpointing the database when using a DatabasePool connection. Checkpointing basically functioned identically to closing the connection as far as the db version behavior. 

As a sanity check, I've tested CRSQLite extension with [SQLite.swift](https://github.com/stephencelis/SQLite.swift) using a similar set up in iOS and the version increments correct on each transaction (ex. an insertion to a table). This makes me think that the way GRDB handles connections somehow prevents the extension from realizing there are new transactions committed. Additionally, this also functions correctly when running the same extension from the SQLite CLI.  

Any advice on how better to understand the details of GRDB's SQLite connections and any pointers to solutions would be greatly appreciated. I'm diving deep into the docs and haven't found a good lead quite yet. Thank you!

**Comment by michaelmedellin** [comment]
I think this could possibly be related to CRSQLite requiring exclusive use of the `commit`, `update`, and `rollback` hooks.

See here: https://github.com/vlcn-io/cr-sqlite/issues/396#issuecomment-1779946129

**Reply by groue** [reply]
Hello @michaelmedellin,

Indeed, CRSQLite [installs](https://github.com/vlcn-io/cr-sqlite/blob/v0.16.3/core/src/crsqlite.c#L100) a [commit hook](https://github.com/vlcn-io/cr-sqlite/blob/v0.16.3/core/src/crsqlite.c#L35) that updates a version number which is [stored in the database file](https://github.com/vlcn-io/cr-sqlite/blob/v0.16.3/core/rs/core/src/bootstrap.rs#L179-L194).

GRDB also [installs](https://github.com/groue/GRDB.swift/blob/v6.29.2/GRDB/Core/TransactionObserver.swift#L690) a [commit hook](https://github.com/groue/GRDB.swift/blob/v6.29.2/GRDB/Core/TransactionObserver.swift#L526) that enables the [database observation features](https://swiftpackageindex.com/groue/grdb.swift/v6.29.2/documentation/grdb/databaseobservation).

The last installed hook wins. Generally, as you say, both libs require exclusive use of all commit, update and rollback hooks.

GRDB does not expose any API that would allow hooks to coexist. And I don't know if CRSQLite does. Hook coexistence is a little bit complicated because the commit hook can [convert a commit into a rollback](https://www.sqlite.org/c3ref/commit_hook.html), a feature used by GRDB. _Chaining_ commit hooks is the difficult part.

I'm not sure of the way forward, from your point of view.

- GRDB ships the low-level [TransactionObserver](https://swiftpackageindex.com/groue/grdb.swift/documentation/grdb/transactionobserver) protocol, in which you may be able to reproduce the CRSQLite features you need.
- It may also be interesting to disable database observation in GRDB, in which case CRSQLite should work.

**Reply by michaelmedellin** [reply]
@groue yep thank you so much for confirming. I don't think there are good approaches for coexisting hooks at the moment — an area for future exploration on my side. 

Re: disabling database observation in GRDB, is there a straight forward approach for doing so?

**Reply by groue** [reply]
> Re: disabling database observation in GRDB, is there a straight forward approach for doing so?

Observation is already disabled in read-only connections, so this should not be too hard. Check what happens when you comment out [those lines](https://github.com/groue/GRDB.swift/blob/v6.29.2/GRDB/Core/Database.swift#L375-L379). Probably CRSQLite will just kick in.

If this works, we could move on to a real pull request, which introduces a new `Configuration.transactionObservationIsEnabled` flag. We'd discuss about the relevant tests.

Does it sound like a plan to you?

**Reply by michaelmedellin** [reply]
@groue yep, commenting those lines worked. Adding this as a supported configuration sounds like a plan.

**Reply by groue** [reply]
Good news! 🥳

I won't ask you for a pull request, because I don't have a clear idea of my requirements, here ;)

Can you live with your fork for a while?

This discussion can of course go on with our ideas.

**Reply by michaelmedellin** [reply]
I can definitely live with my fork for now.  Glad I'm able to still use GRDB. 

An area I'm exploring is if I can live with just using update hooks for observability. It's not ideal, but may work in a pinch.

**Reply by groue** [reply]
> I can definitely live with my fork for now.  Glad I'm able to still use GRDB. 

Cool 👍 

> An area I'm exploring is if I can live with just using update hooks for observability. It's not ideal, but may work in a pinch.

Without the commit and rollback hooks, you can not know if the changes notified to the update hook are committed, or rollbacked. You can rollback transactions if you want, and GRDB automatically rollbacks transactions that throw an unhandled error.

The savepoint stack further complexifies the landscape. To address this, GRDB buffers changes notified to the update hook until the savepoint stack becomes empty. For an illustration, see https://swiftpackageindex.com/groue/grdb.swift/v6.29.2/documentation/grdb/transactionobserver#Database-Changes-And-Transactions
